lib/Dialect/SV/Transforms/HWMemSimImpl.cpp

```diff
@@ -576,36 +559,55 @@ void HWMemSimImpl::generateMemory(HWModuleOp op, FirMemory mem) {
         }
       });
     }
-
+    auto randomMemReg = b.create<sv::RegOp>(
+        b.getIntegerType(llvm::divideCeil(mem.dataWidth, randomWidth) *
+                         randomWidth),
+        b.getStringAttr("_RANDOM_MEM"));
     b.create<sv::InitialOp>([&]() {
       b.create<sv::VerbatimOp>("`INIT_RANDOM_PROLOG_");
 
       // Memory randomization logic.  The entire memory is randomized.
       if (!disableMemRandomization) {
         b.create<sv::IfDefProceduralOp>("RANDOMIZE_MEM_INIT", [&]() {
-          std::string verbatimForLoop;
-          llvm::raw_string_ostream s(verbatimForLoop);
-          s << "for (" << initvar << " = 0; " << initvar << " < " << mem.depth
-            << "; " << initvar << " = " << initvar << " + 1) begin\n"
-            << "  {{0}} = ";
-          auto repetitionCount = llvm::divideCeil(mem.dataWidth, randomWidth);
-          if (repetitionCount > 1)
-            s << "{";
-          for (size_t i = 0; i != repetitionCount; ++i) {
-            if (i > 0)
-              s << ", ";
-            s << "{`RANDOM}";
-          }
-          if (repetitionCount > 1)
-            s << "}";
-          s << ";\n";
-          s << "  Memory[" << initvar << "] = "
-            << "{{0}}[" << mem.dataWidth - 1 << ":" << 0 << "];\n"
-            << "end";
-          b.create<sv::VerbatimOp>(
-              verbatimForLoop, ValueRange{},
-              b.getArrayAttr({hw::InnerRefAttr::get(
-                  op.getNameAttr(), randomMemReg.getInnerSymAttr())}));
+          auto outerLoopIndVarType =
+              b.getIntegerType(llvm::Log2_64_Ceil(mem.depth + 1));
+          auto innerUpperBoundWidth = randomMemReg.getType()
+                                          .getElementType()
+                                          .cast<IntegerType>()
+                                          .getWidth();
+          auto innerLoopIndVarType =
+              b.getIntegerType(llvm::Log2_64_Ceil(innerUpperBoundWidth + 1));
+          // Construct the following nested for loops:
+          // ```
+          //   for (int i = 0; i < mem.depth; i++) begin
+          //     for (int j = 0; j < randomMeg.size; j += 32)
+          //       randomMem[j+31:j] = `RANDOM
+          //     Memory[i] = randomMem[mem.dataWidth - 1: 0];
+          // ```
+          b.create<sv::ForOp>(
+              0, mem.depth, 1, outerLoopIndVarType, "i",
+              [&](BlockArgument outerIndVar) {
```

Comment:
This is so much cleaner than the verbatim stuff above! SV is really getting closer to being an AST dialect 🎉